# SparkSQL and MLlib Assignment

## Part B: MLlib

#### Course: Big Data Systems and Architectures           Master: Business Analytics - Full Time
#### **Student Name**: *Lakkas - Pyknis Evangelos*             **Student ID**:*f2822306*                      **email**: eva.lakkaspyknis@aueb.gr  

In [390]:
import pandas as pd

In [391]:
from pyspark.sql import SparkSession

Creating a Spark session and loading the dataset

In [392]:
spark = SparkSession.builder \
    .appName("IMDB Analysis") \
    .getOrCreate()


In [393]:
imdb_df = spark.read.json("movie.json")

In [394]:
pandas_imdb = imdb_df.toPandas()

In [395]:
pandas_imdb.columns

Index(['_corrupt_record', 'actors', 'countries', 'description', 'directors',
       'genre', 'imdb_url', 'img_url', 'languages', 'metascore', 'rating',
       'runtime', 'tagline', 'title', 'users_rating', 'votes', 'year'],
      dtype='object')

Inspecting the dataset quickly to get familiarized with the format and the values of the initial variables

In [134]:
pandas_imdb.head(8)

_corrupt_record                                             actors  \
0               [                                               None   
1            None  [Timothée Chalamet, Elle Fanning, Liev Schreib...   
2            None  [Emilia Clarke, Hadley Fraser, Sylvia Panacion...   
3            None  [Matthew Broderick, Alan Ruck, Mia Sara, Jeffr...   
4            None  [Robert De Niro, Nick Nolte, Jessica Lange, Ju...   
5            None  [Lindsay Lohan, Rachel McAdams, Tina Fey, Tim ...   
6            None  [Sophia Lillis, Samuel Leakey, Alice Krige, Je...   
7            None  [Chris Pratt, Bryce Dallas Howard, Rafe Spall,...   

                              countries  \
0                                  None   
1                                 [USA]   
2                                 [USA]   
3                                 [USA]   
4                                 [USA]   
5                         [USA, Canada]   
6  [Canada, Ireland, USA, South Africa]   
7                                 [USA]   

                                         description  \
0                                               None   
1  A young couple arrives in New York for a weeke...   
2  Creepy, terrifying chapters from our book of h...   
3  A high school wise guy is determined to have a...   
4  A convicted rapist, released from prison after...   
5  Cady Heron is a hit with The Plastics, the A-l...   
6  A long time ago in a distant fairy tale countr...   
7  When the island's dormant volcano begins roari...   

                                           directors  \
0                                               None   
1                                      [Woody Allen]   
2  [Michael Escobedo, Kelly Hallmark, 4 more cred...   
3                                      [John Hughes]   
4                                  [Martin Scorsese]   
5                                      [Mark Waters]   
6                                       [Oz Perkins]   
7                                      [J.A. Bayona]   

                                  genre  \
0                                  None   
1                     [Comedy, Romance]   
2                    [Horror, Thriller]   
3                              [Comedy]   
4                     [Crime, Thriller]   
5                              [Comedy]   
6  [Fantasy, Horror, Mystery, Thriller]   
7           [Action, Adventure, Sci-Fi]   

                                 imdb_url  \
0                                    None   
1   https://www.imdb.com/title/tt7139936/   
2  https://www.imdb.com/title/tt12384178/   
3   https://www.imdb.com/title/tt0091042/   
4   https://www.imdb.com/title/tt0101540/   
5   https://www.imdb.com/title/tt0377092/   
6   https://www.imdb.com/title/tt9086228/   
7   https://www.imdb.com/title/tt4881806/   

                                             img_url  \
0                                               None   
1  https://m.media-amazon.com/images/M/MV5BODAwZD...   
2  https://m.media-amazon.com/images/M/MV5BMDI5ZD...   
3  https://m.media-amazon.com/images/M/MV5BMDA0Nj...   
4  https://m.media-amazon.com/images/M/MV5BNjdhNz...   
5  https://m.media-amazon.com/images/M/MV5BMjE1MD...   
6  https://m.media-amazon.com/images/M/MV5BOTIyYW...   
7  https://m.media-amazon.com/images/M/MV5BNzIxMj...   

                                languages metascore rating  runtime  \
0                                    None      None   None     None   
1                               [English]        44  PG-13   92 min   
2                               [English]      None     18   91 min   
3                       [English, German]        61  PG-13  103 min   
4                               [English]        73      R  128 min   
5  [English, German, Vietnamese, Swahili]        66  PG-13   97 min   
6                               [English]        64  PG-13   87 min   
7                      [English, Russian]        51  PG-13  128 min   

                

In [135]:
pandas_imdb.tail(20)

_corrupt_record                                             actors  \
62040            None  [Ziggie Bruhn, Meredith Grau, Cameron Jappe, J...   
62041            None  [Kentoku Aoyama, Bethany Butler, Bob Charnley,...   
62042            None  [Zena Keefe, Lionel Belmore, Paul Doucet, Joyc...   
62043            None  [Chris Pratt, Paula Rester, Brandi Jo Perkins,...   
62044            None  [Linda Gillen, Craig Wasson, Louise Hoven, Sal...   
62045            None  [Joe Kirkwood Jr., Cathy Downs, James Gleason,...   
62046            None  [Lauren Chavez-Myers, Israel Wright, Merritt C...   
62047            None  [Matt Nolan, Tami-Adrian George, Reedy Gibbs, ...   
62048            None  [Keith Myers, Tracy Viera, Michael Thyer, Vale...   
62049            None  [Trenda Baldridge, Melissa Barette, Donna Barn...   
62050            None  [David Lee Garver, R.J. Reed, Melissa Hickey, ...   
62051            None  [Vivian Jimenez Hall, Richard D. Curtin, Marga...   
62052            None  [Rene Adams, Alexandra Berardi, Nathan Caldwel...   
62053            None  [Linda Baxter Hardy, James O'Brien, Diane Mela...   
62054            None  [Lash La Rue, Al St. John, Jennifer Holt, Geor...   
62055            None  [Robert Harriell, Cassandra Cooke, Paul Weissm...   
62056            None  [Johnny Mack Brown, Max Terhune, Bill Kennedy,...   
62057            None  [Charlisa Anderson, Emily Bergmann, Jeremiah C...   
62058            None  [Bryan Price, Christa Nicole Wells, Brian Sher...   
62059               ]                                               None   

      countries                                        description  \
62040     [USA]  An ambitious ghost hunter is hired by a wealth...   
62041     [USA]  Two families, one from Japan and one from a sm...   
62042     [USA]                                               None   
62043     [USA]  A man looking for love, a woman looking for en...   
62044     [USA]  Young Mildred, a carhop waitress at O'Rear's D...   
62045     [USA]  Joe and Anne Palooka, along with Joe's trainer...   
62046     [USA]  Missing You is the comical story of Laura, an ...   
62047     [USA]  After two and a half years abroad, Taylor MacL...   
62048     [USA]  The characters. KEVIN has a formula: he earns ...   
62049     [USA]  A girl struggling with bulimia checks into a h...   
62050     [USA]  Guy, far too intelligent for the small town he...   
62051     [USA]  Triangle Woman takes you on a narrative odysse...   
62052     [USA]                                               None   
62053     [USA]  Jeff Coukous thought he had a nice, normal lif...   
62054     [USA]  Taylor's men are robbing incoming supply wagon...   
62055     [USA]  Through a series of String equations, physicis...   
62056     [USA]  With the railroad coming, Nixon is after the r...   
62057     [USA]  The mysterious area of woods known to the loca...   
62058     [USA]  False advertising lures a charming bachelor in...   
62059      None                                               None   

                     directors                          genre  \
62040           [Gene Shibuya]             [Horror, Thriller]   
62041              [Mie Smith]                [Comedy, Drama]   
62042          [John W. Noble]               [Drama, History]   
62043            [Paul Gordon]                [Comedy, Drama]   
62044          [Robert Mandel]                        [Drama]   
62045       [Reginald Le Borg]  [Comedy, Crime, Drama, Sport]   
62046             [Phil Duran]              [Comedy, Romance]   
62047        [Andrew Humeniuk]                        [Drama]   
62048         [Nacho Manubens]                       [Comedy]   
62049        [Kara Kwiecinski]                        [Drama]   
62050            [Marc Berlin]                       [Comedy]   
62051          [Chris Rentzel]                       [Comedy]   
62052         [Nancy Tunstall]                        [Drama]   
62053          [Michael Legg

Checking for the columns data types

In [396]:
print(pandas_imdb.dtypes)

_corrupt_record    object
actors             object
countries          object
description        object
directors          object
genre              object
imdb_url           object
img_url            object
languages          object
metascore          object
rating             object
runtime            object
tagline            object
title              object
users_rating       object
votes              object
year               object
dtype: object


In [397]:
pandas_imdb['runtime'].describe()

count      49893
unique       290
top       90 min
freq        3042
Name: runtime, dtype: object

In [398]:
pandas_imdb['users_rating'].dtype

dtype('O')

In [399]:
import pyspark.sql.functions as F
from pyspark.sql.types import FloatType

Runtime variable has the abbrevation "min" which stands for minutes next to the number which declares the duration of the movie. To make this variable 
a useful input for my model I built a function which removes the commas and the "min" expression and replaces them with a space. After that I setted the data type to float

In [400]:
def clean_runtime(value):
    if value is not None:
        value = value.replace(",", "")
        return float(value.replace("min", ""))
    else:
        return None

In [401]:
clean_runtime_udf = F.udf(clean_runtime, FloatType())

In [402]:
imdb_df = imdb_df.withColumn("runtime_cleaned", clean_runtime_udf("runtime"))

Similarly for genre variable I extracted only the first word from the corrensponding field

In [403]:
imdb_df = imdb_df.withColumn("genre_str", F.concat_ws(", ", "genre"))

In [404]:
imdb_df = imdb_df.withColumn("first_genre", F.split(imdb_df["genre_str"], ", ")[0])

In [405]:
imdb_df = imdb_df.drop("genre_str")

I transformed imdb_df dataset into a pandas dataframe once again in order to update the pandas dataframe with the clean columns. Pandas dataframe is 
more useful to inspect the values of the variables as it is easily readable

In [406]:
pandas_imdb_clean = imdb_df.toPandas()

In [407]:
pandas_imdb_clean.columns

Index(['_corrupt_record', 'actors', 'countries', 'description', 'directors',
       'genre', 'imdb_url', 'img_url', 'languages', 'metascore', 'rating',
       'runtime', 'tagline', 'title', 'users_rating', 'votes', 'year',
       'runtime_cleaned', 'first_genre'],
      dtype='object')

In [408]:
pandas_imdb_clean.head(5)

_corrupt_record                                             actors  \
0               [                                               None   
1            None  [Timothée Chalamet, Elle Fanning, Liev Schreib...   
2            None  [Emilia Clarke, Hadley Fraser, Sylvia Panacion...   
3            None  [Matthew Broderick, Alan Ruck, Mia Sara, Jeffr...   
4            None  [Robert De Niro, Nick Nolte, Jessica Lange, Ju...   

  countries                                        description  \
0      None                                               None   
1     [USA]  A young couple arrives in New York for a weeke...   
2     [USA]  Creepy, terrifying chapters from our book of h...   
3     [USA]  A high school wise guy is determined to have a...   
4     [USA]  A convicted rapist, released from prison after...   

                                           directors               genre  \
0                                               None                None   
1                                      [Woody Allen]   [Comedy, Romance]   
2  [Michael Escobedo, Kelly Hallmark, 4 more cred...  [Horror, Thriller]   
3                                      [John Hughes]            [Comedy]   
4                                  [Martin Scorsese]   [Crime, Thriller]   

                                 imdb_url  \
0                                    None   
1   https://www.imdb.com/title/tt7139936/   
2  https://www.imdb.com/title/tt12384178/   
3   https://www.imdb.com/title/tt0091042/   
4   https://www.imdb.com/title/tt0101540/   

                                             img_url          languages  \
0                                               None               None   
1  https://m.media-amazon.com/images/M/MV5BODAwZD...          [English]   
2  https://m.media-amazon.com/images/M/MV5BMDI5ZD...          [English]   
3  https://m.media-amazon.com/images/M/MV5BMDA0Nj...  [English, German]   
4  https://m.media-amazon.com/images/M/MV5BNjdhNz...          [English]   

  metascore rating  runtime  \
0      None   None     None   
1        44  PG-13   92 min   
2      None     18   91 min   
3        61  PG-13  103 min   
4        73      R  128 min   

                                             tagline  \
0                                               None   
1                                               None   
2       When it comes to murder, they wrote the book   
3    Because life is too beautiful a thing to waste.   
4  There is nothing in the dark that isn't there ...   

                      title users_rating    votes  year  runtime_cleaned  \
0                      None         None     None  None              NaN   
1   A Rainy Day in New York          6.6   21,903  2019             92.0   
2             Murder Manual          2.4      192  2020             91.0   
3  Ferris Bueller's Day Off          7.8  308,847  1986            103.0   
4                 Cape Fear          7.3  165,628  1991            128.0   

  first_genre  
0              
1      Comedy  
2      Horror  
3      Comedy  
4       Crime

Next I searched for the distinct values of the languages variable as we can see the output is pretty chaotic and cannot be a good input for my model because the languages categorical variable will have too many levels which probably will not add much information in the model. Also I couldn't think a transformation that could keep the information of this column in a useful way and with the data type remaining string.

However based on the values of this variable I created a new variable that counts in how many different languages each movie has been translated that
variable may lose some information but it can be a useful input for my model. I named this variable languages count.

In [410]:
distinct_languages = imdb_df.select("languages").distinct().rdd.flatMap(lambda x: x).collect()
print("Distinct values in the languages field:")
for lang in distinct_languages:
    print(lang)

Distinct values in the languages field:
['English', 'Portuguese', 'Finnish']
['English', 'Romanian', 'Russian']
['English', 'Latin', 'French']
['English', 'French', 'German', 'Italian', 'Japanese']
['English', 'Cantonese', 'Japanese', 'Arabic']
['Spanish', 'English']
['English', 'Esperanto']
['English', 'Italian', 'Russian']
['English', 'French', 'Italian', 'Hindi']
['English', 'Russian', 'German', 'French', 'Finnish', 'Italian']
['English', 'French', 'Czech']
['English', 'Russian', 'Spanish']
['Hungarian', 'English']
['English', 'Urdu', 'Arabic', 'Persian', 'French', 'Mandarin']
['English', 'Cantonese', 'Korean']
['English', 'Italian', 'French', 'Japanese']
['English', 'German', 'Italian', 'Japanese']
['English', 'French', 'Spanish']
['English', 'Spanish', 'Portuguese', 'Mandarin', 'Tagalog', 'Russian']
['English', 'Swahili']
['English', 'German', 'Russian']
['English', 'Tagalog']
['English', 'Hawaiian', 'Mandarin', 'None']
['Czech', 'English']
['English', 'Algonquin']
['English', 'Fr

In [411]:
from pyspark.sql.types import StringType
from pyspark.sql.functions import udf

In [412]:
from pyspark.sql.types import IntegerType

Building a function that counts the number of the languages in the languages field based on the length of the list

In [413]:
def count_languages(languages):
    if languages is not None:
        return len(languages)
    else:
        return 0

Creating the languages_count variable and setting it to Integer data type

In [414]:
count_languages_udf = udf(count_languages, IntegerType())

In [415]:
imdb_df = imdb_df.withColumn("languages_count", count_languages_udf("languages"))

Updating the pandas dataframe once again

In [416]:
pandas_imdb_clean = imdb_df.toPandas()

Checking that everything went well

In [417]:
pandas_imdb_clean.head(5)

_corrupt_record                                             actors  \
0               [                                               None   
1            None  [Timothée Chalamet, Elle Fanning, Liev Schreib...   
2            None  [Emilia Clarke, Hadley Fraser, Sylvia Panacion...   
3            None  [Matthew Broderick, Alan Ruck, Mia Sara, Jeffr...   
4            None  [Robert De Niro, Nick Nolte, Jessica Lange, Ju...   

  countries                                        description  \
0      None                                               None   
1     [USA]  A young couple arrives in New York for a weeke...   
2     [USA]  Creepy, terrifying chapters from our book of h...   
3     [USA]  A high school wise guy is determined to have a...   
4     [USA]  A convicted rapist, released from prison after...   

                                           directors               genre  \
0                                               None                None   
1                                      [Woody Allen]   [Comedy, Romance]   
2  [Michael Escobedo, Kelly Hallmark, 4 more cred...  [Horror, Thriller]   
3                                      [John Hughes]            [Comedy]   
4                                  [Martin Scorsese]   [Crime, Thriller]   

                                 imdb_url  \
0                                    None   
1   https://www.imdb.com/title/tt7139936/   
2  https://www.imdb.com/title/tt12384178/   
3   https://www.imdb.com/title/tt0091042/   
4   https://www.imdb.com/title/tt0101540/   

                                             img_url          languages  \
0                                               None               None   
1  https://m.media-amazon.com/images/M/MV5BODAwZD...          [English]   
2  https://m.media-amazon.com/images/M/MV5BMDI5ZD...          [English]   
3  https://m.media-amazon.com/images/M/MV5BMDA0Nj...  [English, German]   
4  https://m.media-amazon.com/images/M/MV5BNjdhNz...          [English]   

  metascore rating  runtime  \
0      None   None     None   
1        44  PG-13   92 min   
2      None     18   91 min   
3        61  PG-13  103 min   
4        73      R  128 min   

                                             tagline  \
0                                               None   
1                                               None   
2       When it comes to murder, they wrote the book   
3    Because life is too beautiful a thing to waste.   
4  There is nothing in the dark that isn't there ...   

                      title users_rating    votes  year  runtime_cleaned  \
0                      None         None     None  None              NaN   
1   A Rainy Day in New York          6.6   21,903  2019             92.0   
2             Murder Manual          2.4      192  2020             91.0   
3  Ferris Bueller's Day Off          7.8  308,847  1986            103.0   
4                 Cape Fear          7.3  165,628  1991            128.0   

  first_genre  languages_count  
0                            0  
1      Comedy                1  
2      Horror                1  
3      Comedy                2  
4       Crime                1

In [418]:
pandas_imdb_clean.columns

Index(['_corrupt_record', 'actors', 'countries', 'description', 'directors',
       'genre', 'imdb_url', 'img_url', 'languages', 'metascore', 'rating',
       'runtime', 'tagline', 'title', 'users_rating', 'votes', 'year',
       'runtime_cleaned', 'first_genre', 'languages_count'],
      dtype='object')

In [419]:
pandas_imdb_clean['users_rating']

0        None
1         6.6
2         2.4
3         7.8
4         7.3
         ... 
62055     6.9
62056     6.2
62057     6.0
62058     6.5
62059    None
Name: users_rating, Length: 62060, dtype: object

Setting the values of users_rating column to float data type

In [420]:
from pyspark.sql.functions import col, trim
from pyspark.sql.types import FloatType

In [421]:
imdb_df = imdb_df.withColumn("users_rating", col("users_rating").cast(FloatType()))

In [422]:
pandas_imdb_clean = imdb_df.toPandas()

Inspecting the data types of the variables that I want to put in my model

In [423]:
print(pandas_imdb_clean.dtypes)

_corrupt_record     object
actors              object
countries           object
description         object
directors           object
genre               object
imdb_url            object
img_url             object
languages           object
metascore           object
rating              object
runtime             object
tagline             object
title               object
users_rating       float32
votes               object
year                object
runtime_cleaned    float32
first_genre         object
languages_count      int32
dtype: object


Checking the values of the metascore column

In [424]:
pandas_imdb_clean['metascore'].unique()

array([None, '44', '61', '73', '66', '64', '51', '69', '70', '84', '65',
       '78', '45', '76', '63', '72', '88', '68', '54', '82', '85', '80',
       '67', '35', '100', '56', '43', '86', '38', '53', '46', '97', '25',
       '36', '41', '28', '74', '42', '50', '18', '77', '60', '59', '57',
       '52', '87', '62', '95', '47', '33', '75', '12', '34', '30', '79',
       '39', '49', '71', '58', '29', '37', '89', '40', '19', '23', '24',
       '99', '16', '26', '93', '55', '20', '27', '17', '31', '9', '91',
       '32', '83', '13', '8', '48', '21', '81', '22', '15', '1', '90',
       '6', '5', '14', '3', '11', '7', '10', '4', '94', '96', '92', '98'],
      dtype=object)

In [425]:
pandas_imdb_clean['metascore'].head(10)

0    None
1      44
2    None
3      61
4      73
5      66
6      64
7      51
8      69
9      70
Name: metascore, dtype: object

In [426]:
pandas_imdb_clean['metascore'].tail(20)

62040    None
62041    None
62042    None
62043    None
62044    None
62045    None
62046    None
62047    None
62048    None
62049    None
62050    None
62051    None
62052    None
62053    None
62054    None
62055    None
62056    None
62057    None
62058    None
62059    None
Name: metascore, dtype: object

In [427]:
pandas_imdb_clean.tail(15)

_corrupt_record                                             actors  \
62045            None  [Joe Kirkwood Jr., Cathy Downs, James Gleason,...   
62046            None  [Lauren Chavez-Myers, Israel Wright, Merritt C...   
62047            None  [Matt Nolan, Tami-Adrian George, Reedy Gibbs, ...   
62048            None  [Keith Myers, Tracy Viera, Michael Thyer, Vale...   
62049            None  [Trenda Baldridge, Melissa Barette, Donna Barn...   
62050            None  [David Lee Garver, R.J. Reed, Melissa Hickey, ...   
62051            None  [Vivian Jimenez Hall, Richard D. Curtin, Marga...   
62052            None  [Rene Adams, Alexandra Berardi, Nathan Caldwel...   
62053            None  [Linda Baxter Hardy, James O'Brien, Diane Mela...   
62054            None  [Lash La Rue, Al St. John, Jennifer Holt, Geor...   
62055            None  [Robert Harriell, Cassandra Cooke, Paul Weissm...   
62056            None  [Johnny Mack Brown, Max Terhune, Bill Kennedy,...   
62057            None  [Charlisa Anderson, Emily Bergmann, Jeremiah C...   
62058            None  [Bryan Price, Christa Nicole Wells, Brian Sher...   
62059               ]                                               None   

      countries                                        description  \
62045     [USA]  Joe and Anne Palooka, along with Joe's trainer...   
62046     [USA]  Missing You is the comical story of Laura, an ...   
62047     [USA]  After two and a half years abroad, Taylor MacL...   
62048     [USA]  The characters. KEVIN has a formula: he earns ...   
62049     [USA]  A girl struggling with bulimia checks into a h...   
62050     [USA]  Guy, far too intelligent for the small town he...   
62051     [USA]  Triangle Woman takes you on a narrative odysse...   
62052     [USA]                                               None   
62053     [USA]  Jeff Coukous thought he had a nice, normal lif...   
62054     [USA]  Taylor's men are robbing incoming supply wagon...   
62055     [USA]  Through a series of String equations, physicis...   
62056     [USA]  With the railroad coming, Nixon is after the r...   
62057     [USA]  The mysterious area of woods known to the loca...   
62058     [USA]  False advertising lures a charming bachelor in...   
62059      None                                               None   

                     directors                          genre  \
62045       [Reginald Le Borg]  [Comedy, Crime, Drama, Sport]   
62046             [Phil Duran]              [Comedy, Romance]   
62047        [Andrew Humeniuk]                        [Drama]   
62048         [Nacho Manubens]                       [Comedy]   
62049        [Kara Kwiecinski]                        [Drama]   
62050            [Marc Berlin]                       [Comedy]   
62051          [Chris Rentzel]                       [Comedy]   
62052         [Nancy Tunstall]                        [Drama]   
62053          [Michael Legge]              [Comedy, Fantasy]   
62054             [Ray Taylor]   [Action, Adventure, Western]   
62055  [Daniel Ulysses Abella]             [Sci-Fi, Thriller]   
62056             [Ray Taylor]                      [Western]   
62057        [William Chaffin]             [Horror, Thriller]   
62058          [Brett Simmons]              [Comedy, Romance]   
62059                     None                           None   

                                    imdb_url  \
62045  https://www.imdb.com/title/tt0043689/   
62046  https://www.imdb.com/title/tt2243417/   
62047  https://www.imdb.com/title/tt0137229/   
62048  https://www.imdb.com/title/tt0462311/   
62049  https://www.imdb.com/title/tt7532654/   
62050  https://www.imdb.com/title/tt0299512/   
62051  https://www.imdb.com/title/tt0985057/   
62052  https://www.imdb.com/title/tt1124055/   
62053  https://www.imdb.com/title/tt0197636/   
62054  https://www.imdb.com/title/tt0039380/   
62055  https://www.imdb.com/title/tt0475272/   
62056  https://www.imdb.com/title/tt0041584/   
620

I noticed with naked eye that there are several NaN values in the metascore column so I searched how many nulls exist exactly in this column

In [428]:
none_metascore_count = imdb_df.filter((col("metascore") == "None") | (col("metascore").isNull())).count()

In [429]:
print("Number of rows with metascore equal to 'none':", none_metascore_count)


Number of rows with metascore equal to 'none': 52449


In [430]:
record_count = imdb_df.count()

print("Number of records in the dataset:", record_count)


Number of records in the dataset: 62060


I defined explicitly the data type of metascore as Integer

In [432]:
imdb_df = imdb_df.withColumn("metascore", col("metascore").cast(IntegerType()))
pandas_imdb_clean = imdb_df.toPandas()

In [433]:
pandas_imdb_clean['metascore'].dtype

dtype('float64')

In [434]:
pandas_imdb_clean['metascore']

0         NaN
1        44.0
2         NaN
3        61.0
4        73.0
         ... 
62055     NaN
62056     NaN
62057     NaN
62058     NaN
62059     NaN
Name: metascore, Length: 62060, dtype: float64

In [435]:
pandas_imdb_clean.columns

Index(['_corrupt_record', 'actors', 'countries', 'description', 'directors',
       'genre', 'imdb_url', 'img_url', 'languages', 'metascore', 'rating',
       'runtime', 'tagline', 'title', 'users_rating', 'votes', 'year',
       'runtime_cleaned', 'first_genre', 'languages_count'],
      dtype='object')

The dataset contains 52449 rows with null metascore I cannot fill in these observations because metascore is my response variable so I will have to
eliminate them in order to fit the model

In [436]:
imdb_df = imdb_df.filter((col("metascore").isNotNull()))

In [437]:
pandas_imdb_clean = imdb_df.toPandas()

In [438]:
pandas_imdb_clean['users_rating'].isna().sum()

0

In [439]:
pandas_imdb_clean['runtime'].isna().sum()

580

In [440]:
pandas_imdb_clean['first_genre'].isna().sum()

0

In [441]:
pandas_imdb_clean['languages_count'].isna().sum()

0

In [374]:
pandas_imdb_clean.shape

(9611, 20)

In [375]:
pandas_imdb_clean['metascore'].isna().sum()

0

In [376]:
pandas_imdb_clean['metascore'].isnull().sum()

0

In [442]:
pandas_imdb_clean['first_genre'].unique()

array(['Comedy', 'Crime', 'Fantasy', 'Action', 'Drama', 'Adventure',
       'Biography', 'Mystery', 'Animation', 'Horror', 'Thriller',
       'Western', 'Romance', 'Musical', 'War', 'Sci-Fi', 'Family',
       'Music', 'Film-Noir', ''], dtype=object)

There are several null values in the runtime_cleaned variable. I found the median movie runtime and filled the null observations with this value

In [448]:
pandas_imdb_clean['runtime_cleaned'].isnull().sum()

580

In [456]:
median_runtime = imdb_df.selectExpr("percentile_approx(runtime_cleaned, 0.5)").collect()[0][0]

In [457]:
imdb_df = imdb_df.withColumn("runtime_cleaned", F.when(col("runtime_cleaned").isNull(), median_runtime).otherwise(col("runtime_cleaned")))

I created a subset which contains only the metascore and its predictors

In [458]:
predictors = ["metascore", "users_rating", "first_genre", "languages_count", "runtime_cleaned"]
subset_df = imdb_df.select(predictors)

In [459]:
from pyspark.sql.functions import when
imdb_df = imdb_df.withColumn("first_genre", when(imdb_df["first_genre"] == "", "unknown").otherwise(imdb_df["first_genre"]))

In [460]:
pandas_imdb_clean = imdb_df.toPandas()

In [461]:
pandas_imdb_clean['runtime_cleaned'].unique()

array([ 92., 103., 128.,  97.,  87., 143., 126., 149., 170., 102.,  90.,
       116., 161., 115., 140., 124., 109., 108., 135.,  98., 110., 158.,
       105., 113., 101., 142., 117., 162., 131., 122., 100., 114., 112.,
        91., 138.,  93.,  99., 125., 144., 107.,  94., 121.,  95., 130.,
       156.,  96., 154., 132.,  84., 104., 106., 134., 111., 137., 150.,
       129.,  89.,  86., 178., 118.,  88., 127., 119.,  83.,  79., 181.,
       145., 123., 133.,  82.,  77.,  78., 165., 136., 120., 176.,  85.,
       148.,  71.,  80., 179.,  73.,  76.,  72., 319.,  81.,  75.,  70.,
        69.,  67.,  74.,  66.,  48.,  62.,  60.,  68.,  63., 182., 169.,
        50., 153., 141., 139., 147., 177., 160., 163., 167., 146., 180.,
       172., 152., 164., 192., 155., 195., 219.,  64., 188., 171., 201.,
       191., 197., 151., 157., 189., 193., 212., 175., 185., 210., 183.,
       159., 174., 187., 238., 166., 202., 209., 194., 168.])

In [462]:
pandas_imdb_clean.columns

Index(['_corrupt_record', 'actors', 'countries', 'description', 'directors',
       'genre', 'imdb_url', 'img_url', 'languages', 'metascore', 'rating',
       'runtime', 'tagline', 'title', 'users_rating', 'votes', 'year',
       'runtime_cleaned', 'first_genre', 'languages_count'],
      dtype='object')

In [463]:
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.regression import LinearRegression
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import VectorAssembler, OneHotEncoder, StringIndexer
from pyspark.ml import Pipeline

Since genre is a string I must encode it in order to create a categorical variable with multiple levels in my model

In [464]:
indexer = StringIndexer(inputCol="first_genre", outputCol="first_genre_indexed")

In [465]:
encoder = OneHotEncoder(inputCols=["first_genre_indexed"], outputCols=["first_genre_encoded"])

In [466]:
assembler = VectorAssembler(inputCols=["users_rating", "languages_count", "first_genre_encoded", 'runtime_cleaned'], outputCol="features")

Creating and Fitting a pipeline which contains the genre encoding and creation of the matrix with the values of the predictors of the metascore

In [467]:
pipeline = Pipeline(stages=[indexer, encoder, assembler])

In [468]:
data = pipeline.fit(subset_df).transform(subset_df)

In [469]:
data = data.select("features", "metascore")

Splitting the data into training and testing datasets

In [470]:
train_data, test_data = data.randomSplit([0.75, 0.25], seed=123)

Defining the regression model and fitting into the training dataset

In [471]:
regression = LinearRegression(featuresCol="features", labelCol="metascore")

In [472]:
model = regression.fit(train_data)

Fitting the model into the testing dataset

In [473]:
predictions = model.transform(test_data)

In [474]:
actual_predicted = predictions.select("metascore", "prediction")

Calculating the r-squared statistic for evaluating the predictions that were made from the linear regression model

In [475]:
evaluator = RegressionEvaluator(labelCol="metascore", predictionCol="prediction", metricName="r2")
r_squared = evaluator.evaluate(predictions)

R-squared is 45% which means that the model is mediocre and it is not the best statistical or machine learning model that can be defined

In [476]:
print("R-squared:", r_squared)

R-squared: 0.4564194939523927


In [477]:
spark.stop()